src/main/scala/uk/gov/nationalarchives/consignmentexport/Main.scala

```diff
@@ -20,37 +23,56 @@ object Main extends CommandIOApp("tdr-consignment-export", "Exports tdr files in
+          rootLocation = config.efs.rootLocation
+          exportId = UUID.randomUUID
+          basePath = s"$rootLocation/$exportId"
+          tarPath = s"$basePath/$consignmentId.tar.gz"
+          bashCommands = BashCommands()
+          graphQlApi = GraphQlApi(config.api.url, config.auth.url)
+          keycloakClient = KeycloakClient(config)
+          s3Files = S3Files(S3Utils(s3Async))
+          bagit = Bagit()
+          validator = Validator(consignmentId)
+          //Export datetime generated as value needed in bag metadata and DB table
+          //Cannot use the value from DB table in bag metadata, as bag metadata created before bagging
+          //and cannot update DB until bag creation successfully completed
+          exportDatetime = ZonedDateTime.now(ZoneOffset.UTC)
+          consignmentResult <- graphQlApi.getConsignmentMetadata(config, consignmentId)
+          consignmentData <- IO.fromEither(validator.validateConsignmentResult(consignmentResult))
+          _ <- IO.fromEither(validator.validateConsignmentHasFiles(consignmentData))
+          bagMetadata <- BagMetadata(keycloakClient).generateMetadata(consignmentId, consignmentData, exportDatetime)
+          validatedFileMetadata <- IO.fromEither(validator.extractFileMetadata(consignmentData.files))
+          _ <- s3Files.downloadFiles(validatedFileMetadata, config.s3.cleanBucket, consignmentId, basePath)
+          bag <- bagit.createBag(consignmentId, basePath, bagMetadata)
+          fileMetadataCsv <- BagAdditionalFiles(bag.getRootDir).createFileMetadataCsv(validatedFileMetadata)
+          checksums <- ChecksumCalculator().calculateChecksums(fileMetadataCsv)
+          _ <- bagit.writeTagManifestRows(bag, checksums)
+          // The owner and group in the below command have no effect on the file permissions. It just makes tar idempotent
+          _ <- bashCommands.runCommand(s"tar --sort=name --owner=root:0 --group=root:0 --mtime ${java.time.LocalDate.now.toString} -C $basePath -c ./$consignmentId | gzip -n > $tarPath")
+          _ <- bashCommands.runCommand(s"sha256sum $tarPath > $tarPath.sha256")
+          _ <- s3Files.uploadFiles(config.s3.outputBucket, consignmentId, tarPath)
+          _ <- graphQlApi.updateExportLocation(config, consignmentId, s"s3://${config.s3.outputBucket}/$consignmentId.tar.gz", exportDatetime)
+          _ <- stepFunction.publishSuccess(taskToken,
+            ExportOutput(consignmentData.userid,
+              bagMetadata.get(InternalSenderIdentifierKey).get(0),
+              bagMetadata.get(SourceOrganisationKey).get(0)))
+        } yield ExitCode.Success
+
+        exitCode.recoverWith({
+          case rte: RuntimeException =>
+            for {
+              _ <- stepFunction.publishFailure(taskToken, s"$exportFailedErrorMessage: ${rte.getMessage}")
+            } yield throw rte
+          case _ =>
+            for {
+              _ <- stepFunction.publishFailure(taskToken, exportFailedErrorMessage)
```

Comment:
The idea was to ensure that the publishFailure is always called, regardless of why the export process failed, ie a default

Reply:
I wonder if this could be done a little simpler without chucking the exceptions
```
exportOps.map {
      case FileExport(consignmentId) =>
         val exitCode = for {
           config <- config()
           //normal export stuff
           _ <- IO()//send success to step function
         } yield ExitCode.Success
         exitCode.handleErrorWith(f => {
           //Send failure to step function
           IO.raiseError(f)
         })
    }
```
I think this would also work when the task token is mandatory and you'll stick get the stack trace of whatever was up with it in the logs.

Reply:
Thanks. Have refactored along these lines

Reply:
This looks good to me, thanks. Sounds like we're letting the IO command line framework deal with the error sensibly. 👍